### ISSUE TITLE
Planner delete with "dedup" semi-join on replicate table may cause ERROR

### ISSUE THREAD

**Issue by wuyuhao28** [greenplum-db/gpdb#17460]
## Bug Report
After #9703 we support "dedup" semi-join, but with this Path in planner, delete on replicated table may cause ERROR.
ERROR message would be like:
_distribution key of the tuple (0, 1) doesn't belong to current segment (actually from seg1) (nodeModifyTable.c:867)_

### Greenplum version or build
7X with main branch

### Step to reproduce the behavior
```
create table t (a int ) distributed replicated;
create table foo (a int);
insert into t values(1);
insert into foo select i from generate_series(1, 10000) i;
analyze t;
analyze foo;
delete from t where a in (select a from foo);
```

From the plan we can see that dedup semi-join and Delete has different locus，which is obviously wrong.
```
                                                   QUERY PLAN
-----------------------------------------------------------------------------------------------------------------
 Delete on public.t  (cost=46.85..46.90 rows=4 width=16)
   ->  HashAggregate  (cost=46.85..46.90 rows=4 width=16)
         Output: t.ctid, t.gp_segment_id, foo.ctid
         Group Key: (RowIdExpr)
         ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=1.04..46.84 rows=4 width=16)
               Output: t.ctid, t.gp_segment_id, foo.ctid, (RowIdExpr)
               Hash Key: (RowIdExpr)
               ->  Hash Join  (cost=1.04..46.75 rows=4 width=16)
                     Output: t.ctid, t.gp_segment_id, foo.ctid, (RowIdExpr)
                     Hash Cond: (foo.a = t.a)
                     ->  Seq Scan on public.foo  (cost=0.00..37.33 rows=3333 width=10)
                           Output: foo.ctid, foo.a
                     ->  Hash  (cost=1.03..1.03 rows=1 width=14)
                           Output: t.ctid, t.gp_segment_id, t.a, (RowIdExpr)
                           ->  Redistribute Motion 1:3  (slice2; segments: 1)  (cost=0.00..1.03 rows=1 width=14)
                                 Output: t.ctid, t.gp_segment_id, t.a, (RowIdExpr)
                                 Hash Key: t.a
                                 ->  Seq Scan on public.t  (cost=0.00..1.01 rows=1 width=14)
                                       Output: t.ctid, t.gp_segment_id, t.a, RowIdExpr
 Optimizer: Postgres-based planner
(20 rows)
```

**Comment by kainwen** [comment]
Looks like the piece of code has side-effect and will impact later plan's path locus `cdbpath.c:cdbpath_motion_for_join`:

```c
			else if (CdbPathLocus_IsSegmentGeneral(outer.locus))
			{
				CdbPathLocus_MakeSingleQE(&outer.locus,
										  CdbPathLocus_CommonSegments(inner.locus,
																	  outer.locus));
				outer.path->locus = outer.locus;
			}
```

A quick try with the following diff:

```diff
diff --git a/src/backend/cdb/cdbpath.c b/src/backend/cdb/cdbpath.c
index 58e0570d16..b843464fc0 100644
--- a/src/backend/cdb/cdbpath.c
+++ b/src/backend/cdb/cdbpath.c
@@ -1436,6 +1436,9 @@ cdbpath_motion_for_join(PlannerInfo *root,
                         * a risky assumption. And it didn't work on SegmentGeneral paths
                         * (i.e. replicated tables) anyway.
                         */
+                       if (root->upd_del_replicated_table > 0)
+                               goto fail;
+
                        if (!CdbPathLocus_IsPartitioned(inner.locus))
                                goto fail;

@@ -1466,6 +1469,8 @@ cdbpath_motion_for_join(PlannerInfo *root,
                        break;

                case JOIN_DEDUP_SEMI_REVERSE:
+                       if (root->upd_del_replicated_table > 0)
+                               goto fail;
                        /* same as JOIN_DEDUP_SEMI, but with inner and outer reversed */
                        if (!CdbPathLocus_IsPartitioned(outer.locus))
                                goto fail;
```

can generate a correct plan:

```
gpadmin=# explain ( verbose) delete from t where a in (select a from foo);
                                               QUERY PLAN
--------------------------------------------------------------------------------------------------------
 Delete on public.t  (cost=295.67..296.81 rows=13 width=16)
   ->  Hash Semi Join  (cost=295.67..296.81 rows=13 width=16)
         Output: t.ctid, t.gp_segment_id, foo.ctid
         Hash Cond: (t.a = foo.a)
         ->  Seq Scan on public.t  (cost=0.00..1.01 rows=1 width=14)
               Output: t.ctid, t.gp_segment_id, t.a
         ->  Hash  (cost=170.67..170.67 rows=10000 width=10)
               Output: foo.ctid, foo.a
               ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..170.67 rows=10000 width=10)
                     Output: foo.ctid, foo.a
                     ->  Seq Scan on public.foo  (cost=0.00..37.33 rows=3333 width=10)
                           Output: foo.ctid, foo.a
 Optimizer: Postgres-based planner
(13 rows)
```

**Comment by kainwen** [comment]
Plus, `unique row id` skill does not consider the case of `update|delete`'s join of replicated, the above code brute-forcely set the locus to singleQE, which is not good for such case because replicated table's CTID is not consistent across segments.

Besides, one interesting part is gpdb banned direct access `gp_segment_id` field of replicated table from user side, however, GPDB planner contains a code path to use it (like unique_row_id_path)

**Comment by wuyuhao28** [comment]
> Plus, `unique row id` skill does not consider the case of `update|delete`'s join of replicated, the above code brute-forcely set the locus to singleQE, which is not good for such case because replicated table's CTID is not consistent across segments.
> 
> Besides, one interesting part is gpdb banned direct access `gp_segment_id` field of replicated table from user side, however, GPDB planner contains a code path to use it (like unique_row_id_path)

Interesting, this reminds me that if we can forbid dedup semi join when the targetlist contains `gp_segment_id`, then this bug would be solved?

**Comment by kainwen** [comment]
> > Plus, `unique row id` skill does not consider the case of `update|delete`'s join of replicated, the above code brute-forcely set the locus to singleQE, which is not good for such case because replicated table's CTID is not consistent across segments.
> > Besides, one interesting part is gpdb banned direct access `gp_segment_id` field of replicated table from user side, however, GPDB planner contains a code path to use it (like unique_row_id_path)
> 
> Interesting, this reminds me that if we can forbid dedup semi join when the targetlist contains `gp_segment_id`, then this bug would be solved?

Given current implementation of unique_row_id path, my idea now is: do not generate dedup semi join when updating and deleting replicated table.

**Comment by kainwen** [comment]
This is my way to fix https://github.com/greenplum-db/gpdb/pull/17464
@wuyuhao28 please help review and see if there is other good ways.

**Comment by wuyuhao28** [comment]
> This is my way to fix #17464 @wuyuhao28 please help review and see if there is other good ways.

Glad to see your fix, I'll review it soon
